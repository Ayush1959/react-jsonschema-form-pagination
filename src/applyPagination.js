import React, { Component } from "react";
import PropTypes from "prop-types";
import deepequal from "deep-equal";
import formWithTabs from "./FormWithTabs";
import { isDevelopment } from "./utils";
import splitInLayers from "./splitter/splitInLayers";
import Tabs from "./Tabs";

export default function applyPagination(FormComponent, TabComponent = Tabs) {
  const FormWithTabs = formWithTabs(FormComponent, TabComponent);

  class FormWithPagination extends Component {
    constructor(props) {
      super(props);

      let { formData, tabData, schema, uiSchema } = this.props;

      this.formData = formData ? formData : {};

      this.layers = splitInLayers(schema, uiSchema, tabData);
      let activeTabs = [];
      this.layers.updateActiveTabs(activeTabs, 0);
      this.state = { formData, activeTabs };
    }

    sameData = formData => {
      return (
        deepequal(this.props.formData, formData) ||
        deepequal(this.formData, formData)
      );
    };

    sameLayers = nextProps => {
      const toComparable = ({ tabData, schema, uiSchema }) => {
        return { tabData, schema, uiSchema };
      };
      return deepequal(toComparable(nextProps), toComparable(this.props));
    };

    componentWillReceiveProps(nextProps) {
      if (!this.sameLayers(nextProps)) {
        let { schema, uiSchema, tabData } = nextProps;
        this.layers = splitInLayers(schema, uiSchema, tabData);
      }
      if (!this.sameData(nextProps.formData)) {
        this.formData = nextProps.formData;
        this.setState({ formData: nextProps.formData });
      }
    }

    handleTabChange = index => tabID => {
      let activeTabs = this.state.activeTabs.slice(0, index + 1);
      activeTabs[index] = tabID;
      this.layers.updateActiveTabs(activeTabs);
      if (this.props.onTabChange) {
        this.props.onTabChange(activeTabs, this.state.activeTabs);
      }
      this.setState({ activeTabs });
    };

    handleOnChange = state => {
      this.formData = state.formData;
      this.setState({ formData: state.formData });
      if (this.props.onChange) {
        this.props.onChange(state);
      }
    };

    shouldComponentUpdate(nextProps, nextState) {
      let sameActive = deepequal(nextState.activeTabs, this.state.activeTabs);
      if (!sameActive) {
        return true;
      }
      let sameProps = deepequal(
        Object.assign({}, this.props, { formData: {} }),
        Object.assign({}, nextProps, { formData: {} })
      );
      if (!sameProps) {
        return true;
      }
      let sameState = deepequal(
        Object.assign({}, this.state, { formData: this.formData }),
        nextState
      );
      if (!sameState) {
        return true;
      }
      let sameData = this.sameData(nextProps.formData);
      return !sameProps || !sameState || !sameData;
    }

    render() {
      let subForms = this.layers.toSubForms(this.state.activeTabs);
      let formData = this.formData;
      return (
        <div>
          {subForms.map((conf, i) => {
            let allConf = Object.assign({}, this.props, conf, { formData });
            return (
              <FormWithTabs
                key={i}
                {...allConf}
                onChange={this.handleOnChange}
                onTabChange={this.handleTabChange(i)}
              />
            );
          })}
          {this.props.children}
        </div>
      );
    }
  }

  if (isDevelopment()) {
    FormWithPagination.propTypes = {
      schema: PropTypes.shape({
        type: function(props, propName, componentName) {
          if (props[propName] !== "object") {
            return new Error(
              `Only "object" schemas supported by pagination for ${componentName}.`
            );
          }
        },
      }),
      tabData: PropTypes.array,
    };
  }

  return FormWithPagination;
}
