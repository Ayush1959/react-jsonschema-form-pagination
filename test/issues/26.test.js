import React from "react";
import Form from "react-jsonschema-form";
import applyPagination from "../../src";
import renderer from "react-test-renderer";

const schema = {
  type: "object",
  title: "Encounter",
  required: [],
  properties: {
    encounter: { type: "string" },
    firstName: { type: "string" },
    lastName: { type: "string" },
    age: { type: "integer" },
    nickName: { type: "string" },
    bio: { type: "string" },
    password: { type: "string" },
    telephone: { type: "string" },
  },
};

const uiSchema = {
  firstNameAlias: {
    "ui:tabID": ["0", "lastName"],
  },
  lastName: {
    "ui:tabID": ["0", "lastName"],
  },
  age: {
    "ui:tabID": ["0", "firstName", "age"],
  },
  nickName: {
    "ui:tabID": ["0", "firstName", "nickName"],
  },
  bio: {
    "ui:tabID": "1",
  },
  password: {
    "ui:tabID": ["2"],
  },
  telephone: {
    "ui:tabID": "2",
  },
  "ui:tabAlias": {
    firstName: "firstNameAlias",
  },
};

test("encounter title only once", () => {
  let ResForm = applyPagination(Form);
  const component = renderer.create(
    <ResForm schema={schema} uiSchema={uiSchema} />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
