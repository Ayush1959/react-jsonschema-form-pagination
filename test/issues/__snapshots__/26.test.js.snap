// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encounter title only once 1`] = `
<div>
  <div>
    <form
      acceptCharset={undefined}
      action={undefined}
      autoComplete={undefined}
      className="rjsf"
      encType={undefined}
      id={undefined}
      method={undefined}
      name={undefined}
      noValidate={false}
      onSubmit={[Function]}
      target={undefined}
    >
      <div
        className="form-group field field-object"
      >
        <fieldset>
          <legend
            id="root__title"
          >
            Encounter
          </legend>
          <div
            className="form-group field field-string"
          >
            <label
              className="control-label"
              htmlFor="root_encounter"
            >
              encounter
            </label>
            <div />
            <input
              autoFocus={false}
              className="form-control"
              disabled={false}
              id="root_encounter"
              label="encounter"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              readOnly={false}
              required={false}
              type="text"
              value=""
            />
            <div />
            <div />
          </div>
          <div
            className="form-group field field-string"
          >
            <label
              className="control-label"
              htmlFor="root_firstName"
            >
              firstName
            </label>
            <div />
            <input
              autoFocus={false}
              className="form-control"
              disabled={false}
              id="root_firstName"
              label="firstName"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              readOnly={false}
              required={false}
              type="text"
              value=""
            />
            <div />
            <div />
          </div>
        </fieldset>
        <div />
        <div />
      </div>
      <div />
    </form>
    <fieldset>
      <div
        className="form-group col-md-12"
      >
        <ul
          className="nav nav-pills"
        >
          <li
            className="active"
            onClick={[Function]}
          >
            <a>
              0
            </a>
          </li>
          <li
            className={null}
            onClick={[Function]}
          >
            <a>
              1
            </a>
          </li>
          <li
            className={null}
            onClick={[Function]}
          >
            <a>
              2
            </a>
          </li>
        </ul>
      </div>
    </fieldset>
  </div>
  <div>
    <form
      acceptCharset={undefined}
      action={undefined}
      autoComplete={undefined}
      className="rjsf"
      encType={undefined}
      id={undefined}
      method={undefined}
      name={undefined}
      noValidate={false}
      onSubmit={[Function]}
      target={undefined}
    >
      <div
        className="form-group field field-object"
      >
        <fieldset>
          <div
            className="form-group field field-string"
          >
            <label
              className="control-label"
              htmlFor="root_firstName"
            >
              firstName
            </label>
            <div />
            <input
              autoFocus={false}
              className="form-control"
              disabled={false}
              id="root_firstName"
              label="firstName"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              readOnly={false}
              required={false}
              type="text"
              value=""
            />
            <div />
            <div />
          </div>
        </fieldset>
        <div />
        <div />
      </div>
      <div />
    </form>
    <fieldset>
      <div
        className="form-group col-md-12"
      >
        <ul
          className="nav nav-pills"
        >
          <li
            className="active"
            onClick={[Function]}
          >
            <a>
              lastName
            </a>
          </li>
          <li
            className={null}
            onClick={[Function]}
          >
            <a>
              firstName
            </a>
          </li>
        </ul>
      </div>
    </fieldset>
  </div>
  <div>
    <form
      acceptCharset={undefined}
      action={undefined}
      autoComplete={undefined}
      className="rjsf"
      encType={undefined}
      id={undefined}
      method={undefined}
      name={undefined}
      noValidate={false}
      onSubmit={[Function]}
      target={undefined}
    >
      <div
        className="form-group field field-object"
      >
        <fieldset>
          <div
            className="form-group field field-string"
          >
            <label
              className="control-label"
              htmlFor="root_lastName"
            >
              lastName
            </label>
            <div />
            <input
              autoFocus={false}
              className="form-control"
              disabled={false}
              id="root_lastName"
              label="lastName"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              readOnly={false}
              required={false}
              type="text"
              value=""
            />
            <div />
            <div />
          </div>
        </fieldset>
        <div />
        <div />
      </div>
      <div />
    </form>
    <div />
  </div>
</div>
`;
